source/unix/openssl_platform_init.c

```diff
@@ -72,27 +220,144 @@ static int s_resolve_libcrypto_hmac(enum aws_libcrypto_version version, void *mo
-    hmac_ctx_table.reset_fn = reset_fn;
-    hmac_ctx_table.free_fn = free_fn;
-    hmac_ctx_table.init_fn = init_fn;
-    hmac_ctx_table.clean_up_fn = clean_up_fn;
-    hmac_ctx_table.update_fn = update_fn;
-    hmac_ctx_table.final_fn = final_fn;
-    hmac_ctx_table.init_ex_fn = init_ex_fn;
-    g_aws_openssl_hmac_ctx_table = &hmac_ctx_table;
+    if (new_fn) {
+        /* Fill out the vtable for the requested version */
+        hmac_ctx_table.new_fn = new_fn;
+        hmac_ctx_table.reset_fn = reset_fn;
+        hmac_ctx_table.free_fn = free_fn;
+        hmac_ctx_table.init_fn = init_fn;
+        hmac_ctx_table.clean_up_fn = clean_up_fn;
+        hmac_ctx_table.update_fn = update_fn;
+        hmac_ctx_table.final_fn = final_fn;
+        hmac_ctx_table.init_ex_fn = init_ex_fn;
+        g_aws_openssl_hmac_ctx_table = &hmac_ctx_table;
+        return true;
+    }
+#endif
+    return false;
+}
 
-    return version;
+static int s_resolve_libcrypto_hmac(enum aws_libcrypto_version version, void *module) {
+    switch (version) {
+        case AWS_LIBCRYPTO_LC:
+            return s_resolve_hmac_lc(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_1_1_1:
+            return s_resolve_hmac_111(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_1_0_2:
+            return s_resolve_hmac_102(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_NONE:
+            AWS_FATAL_ASSERT(!"Attempted to resolve invalid libcrypto HMAC API version AWS_LIBCRYPTO_NONE");
+    }
+
+    return AWS_LIBCRYPTO_NONE;
 }
 
+/* EVP_MD_CTX API */
+/* 1.0.2 NOTE: these are macros in 1.1.x, so we have to undef them to weak link */
+#if defined(EVP_MD_CTX_create)
+#    pragma push_macro("EVP_MD_CTX_create")
+#    undef EVP_MD_CTX_create
+#endif
+extern EVP_MD_CTX *EVP_MD_CTX_create(void) __attribute__((weak, used));
+static evp_md_ctx_new s_EVP_MD_CTX_create = EVP_MD_CTX_create;
+#if defined(EVP_MD_CTX_create)
+#    pragma pop_macro("EVP_MD_CTX_create")
+#endif
+
+#if defined(EVP_MD_CTX_destroy)
+#    pragma push_macro("EVP_MD_CTX_destroy")
+#    undef EVP_MD_CTX_destroy
+#endif
+extern void EVP_MD_CTX_destroy(EVP_MD_CTX *) __attribute__((weak, used));
+static evp_md_ctx_free s_EVP_MD_CTX_destroy = EVP_MD_CTX_destroy;
+#if defined(EVP_MD_CTX_destroy)
+#    pragma pop_macro("EVP_MD_CTX_destroy")
+#endif
+
 extern EVP_MD_CTX *EVP_MD_CTX_new(void) __attribute__((weak, used));
 extern void EVP_MD_CTX_free(EVP_MD_CTX *) __attribute__((weak, used));
 extern int EVP_DigestInit_ex(EVP_MD_CTX *, const EVP_MD *, ENGINE *) __attribute__((weak, used));
 extern int EVP_DigestUpdate(EVP_MD_CTX *, const void *, size_t) __attribute__((weak, used));
 extern int EVP_DigestFinal_ex(EVP_MD_CTX *, unsigned char *, unsigned int *) __attribute__((weak, used));
 
-static int s_resolve_libcrypto_md(enum aws_libcrypto_version version, void *module) {
+bool s_resolve_md_102(void *module) {
+#if defined(OPENSSL_IS_AWSLC)
```

Comment:
#else or some other split here

Reply:
not needed, if you look, I put the return false outside the #if